lib/systems/src/TorqueControllerMux.cpp

```diff
@@ -85,5 +87,94 @@ void TorqueControllerMux::tick(
         }
 
         drivetrainCommand_ = controllerOutputs_[static_cast<int>(muxMode_)].command;
+
+        applyPowerLimit(&drivetrainCommand_);
+        applyTorqueLimit(&drivetrainCommand_);
+        applyPosSpeedLimit(&drivetrainCommand_);
+
+    }
+}
+
+/*
+    Apply power limit such that the mechanical power of all wheels never
+    exceeds the preset mechanical power limit. Scales all wheels down to
+    preserve functionality of torque controllers
+*/
+void TorqueControllerMux::applyPowerLimit(DrivetrainCommand_s* command)
+{
+    float net_torque_mag = 0;
+    float net_wheelspeed = 0;
+
+    for (int i = 0; i < NUM_MOTORS; i++) {
+        // get the total magnitude of torque from all 4 wheels
+        #ifdef ARDUINO_TEENSY41 // screw arduino.h macros
+        net_torque_mag += abs(command->torqueSetpoints[i]);
+        net_wheelspeed += abs(command->speeds_rpm[i]);
+        #else
+        net_torque_mag += std::abs(command->torqueSetpoints[i]);
+        net_wheelspeed += std::abs(command->speeds_rpm[i]);
+        #endif
+    }
+
+    // get current total mechanical power
+    float cur_power = net_torque_mag * (net_wheelspeed * RPM_TO_RAD_PER_SECOND);
+    
+    // only evaluate power limit if current power exceeds it
+    if (cur_power > (MAX_POWER_LIMIT)) {
+        for (int i = 0; i < NUM_MOTORS; i++) {
+            // calculate the percent of total torque requested per wheel
+            float torque_percent = command->torqueSetpoints[i] / net_torque_mag;
+            // based on the torque percent and max power limit, get the max power each wheel can use
+            float power_per_corner = (torque_percent * MAX_POWER_LIMIT);
```

Comment:
does it make sense to scale power with torque? power is nonlinear

Reply:
Discussed this with Luke in shop, we are going to go ahead with this and monitor the performance, but this shouldn't be too much of a concern since we are only consistently hitting power limits at high speed in straights/accel.